### DISCUSSION TITLE
RFC: Native DNS on top of pubkey-hash addresses

### DISCUSSION THREAD

**Discussion by sebastien-forestier** [massalabs/massa#3454] [Ideas]
### Introduction

We need a _native_ DNS, instead of a smart contract DNS so that you can simply send coins to myaddress.massa without specifying to which DNS contract you refer to. We could hardcode only the DNS smart contract address natively to achieve this, but it's probably a good idea to implement the full DNS natively to save gas as those are a few simple functions used very frequently.

@damip proposed a native DNS (https://github.com/massalabs/massa/discussions/2646) **replacing** pubkey-hash addresses with ".massa" chosen addresses, where ".massa" chosen addresses are associated to pubkeys (not pubkey hash), with no P2PK (pay to pubkey) possibility.

There are several downsides with this approach:
- no more soulbound tokens,
- we can't create an address off-chain,
- requiring the whole crypto industry to change their habits.

Here I instead propose building the native DNS on top of pubkey-hash addresses.

This is a general description, not a detailed specification.

**Note:** With both approaches we gain a small amount of Account Abstraction (AA), but the full AA can only be achieved with smart contract accounts, for which I think the simplest approach for Massa would be to follow [EIP-2938](https://eips.ethereum.org/EIPS/eip-2938) with a new type of transaction and a PAYGAS opcode (cf https://github.com/massalabs/massa/discussions/3456 discussion).

### Address
A domain name is a `.massa string`.
Now, an `address` can be either a `pubkey-hash` or a `.massa string` domain name.

### State
We add a new table `DNS` to the state.
The `DNS` table is composed of `(.massa string, address, is_immutable)` rows, indexed by the first field (whose items appear at most once).

The first field `.massa string` is the domain name that was booked, and which points to the second field, the `address`.
The `is_immutable` field specifies if `address` can change in the future.

### Caller and Login
A new ABI call `LOGIN(dn)` allows to login to a domain name `dn`, changing the current identity/account/caller address to `dn` and `LOGOUT` reverts the caller address to the previous one.

`LOGIN(dn)` succeeds only if the current caller is the address associated to `dn` in the DNS.

A pubkey-hash address can login to a domain name address which can login to another domain name address etc.

### Booking/Transferring/Freezing/Freeing a domain name
A new transaction type `DNS_TX` allows to book, transfer, freeze or free a domain name.

`DNS_TX(BOOK, dn, address)` books the domain name `dn` and assigns it to any address `address`. Succeeds only if `dn` is not already booked.

`DNS_TX(TRANSFER, dn, address)` transfers the domain name `dn` and assigns it to any address `address`. Succeeds only if the transaction caller is the address associated to `dn` in the DNS and the `is_immutable` field is `False`.

`DNS_TX(FREEZE, dn)` freezes the domain name `dn` by setting the `is_immutable` field to `True`. Succeeds only if the transaction caller is the address associated to `dn` in the DNS and the `is_immutable` field is `False`.

`DNS_TX(FREE, dn)` frees the domain name `dn`. Succeeds only if the transaction caller is the address associated to `dn` in the DNS and the `is_immutable` field is `False`.

New ABI calls allow to do the same operations from smart contracts.

### Resolving a domain name
When a normal transaction sends coins to an address that is not a pubkey-hash, then the address is resolved with the DNS (recursively if needed), and the coins are transfered to the pubkey-hash. Succeeds only if a pubkey-hash was resolved, fails if detecting a cycle or if there is no address associated.

A `RESOLVE(address)` ABI allows to resolve an address from a smart contract. Returns the pubkey-hash or fails.

### Storage Costs
Works exactly as for other state tables, paying storage costs when increasing state size, getting back storage costs when decreasing state size.

### Staking
Only pubkey-hash addresses can buy rolls and stake.

### Example Use Case: ERC-20
An ERC-20 contract stores how much coins a given address has.
Now the address can be either a `pubkey-hash` or a `.massa string` domain name.
In order to spend the ERC-20 of a domain name address, one has to `LOGIN` before.

### Comments:
- Should we use a `DNS` table with `(.massa string, owner_address, resolved_address)` to cover more use cases ?
- Should we allow `.massa` addresses to be a key in the ledger to cover more use cases ? 
- Someone can build a marketplace to which we can transfer our domain names and set a minimum price or setup an auction, and others can buy it.

**Comment by damip** [comment]
Caveat: what happens if I'm logged in damir.massa and while being logged in, I change the owner/target or delete the domain damir.massa ?

**Reply by sebastien-forestier** [reply]
It logs you out

**Comment by qdrn** [comment]
Open question: what about the reverse lookup ?
If we limit to 1 address-1 name domain it's easy. Otherwise what do we want to do ?

**Reply by sebastien-forestier** [reply]
In this proposition there is no need for (address -> domain name) lookup nor for limiting to 1 domain name per address

**Reply by qdrn** [reply]
It's not mandatory but it could maybe be useful to add this feature.

**Reply by sebastien-forestier** [reply]
This index+ABI/API could be useful in some situations, e.g. I forgot which domains I booked and I want to ask my node / ask inside a smart contract, but the tradeoff is that it doubles the DNS space usage and so the cost of registration. I don't know what's best

**Comment by AurelienFT** [comment]
I like the proposal. I didn't understand the `login` ABI and his use-cases, especially the difference with `resolve`, one can act as the address (login) and the other only give the information of the address ?

Personal thoughts about the comments : 

> Should we use a DNS table with (.massa string, owner_address, resolved_address) to cover more use cases ?

I don't see a real usage as soon as the address has been transferred we don't need to have the owner informations. It would take some storage space and so cost more coins to create a DNS address for no real value. Maybe worth implementing it in a v2 but I'm not a fan of paying more for something in the future.

> Should we allow .massa addresses to be a key in the ledger to cover more use cases ?

What would be the use cases ? It could add a lot of computation power no ? I don't have real opinions right now

**Reply by sebastien-forestier** [reply]
> I like the proposal. I didn't understand the `login` ABI and his use-cases, especially the difference with `resolve`, one can act as the address (login) and the other only give the information of the address ?

Yes exactly, you login to your account if you want to be identified as `aurelien.massa` for instance to spend ERC20 that you received at `aurelien.massa`.
Now if you want to know the address behind `seb.massa` because you want to send him a soulbound token, you use `RESOLVE(seb.massa)`. The soul is in the key.

**Reply by AurelienFT** [reply]
I don't really understand why we need the login. Because if I receive coins on `aurelien.massa` as the domain name is own by me so when I receive coins, couldn't be possible that they go directly to my address and we never use the `login` method we just resolve before transferring from my address ?
Or you want the domain name to have a special balance etc that are transferred when the domain name is transferred ?

**Reply by damip** [reply]
Login just allows choosing what appears as the caller in the callee's call stack

**Reply by AurelienFT** [reply]
Do you have any use case where it's useful ? I'm having hard time visualizing. For me it's just an "alias" to send coins to a more readable address.

(Maybe use case : have a readable name as a sender ?)

**Reply by damip** [reply]
The callee can then have different access control depending on the logged in caller identity

**Reply by sebastien-forestier** [reply]
On the native side, it's similar to an alias, but on the smart contract side, it's more like account abstraction.
You cannot use your account if you don't login.
If someone sent ERC20 coins to `aurelien.massa`, then the only way to spend it is to login to `aurelien.massa`.

**Reply by qdrn** [reply]
Don't we want to do that at the account abstraction level ?

**Reply by AurelienFT** [reply]
> If someone sent ERC20 coins to aurelien.massa, then the only way to spend it is to login to aurelien.massa.

So the coins are linked to my Domain Name not my account if I transfer my domain name the coins are transferred ?

**Reply by sebastien-forestier** [reply]
The developer of the ERC20 / NFT / soulbound token decides, but the idea is to link ERC20 and NFTs to the domain name, it makes sense to link it to the underlying address only for soulbound tokens (the developer will use `resolve` for this).

**Reply by sebastien-forestier** [reply]
Now, we could also decide to allow domain name keys in the native ledger to allow account abstraction of native coins, but I'm not sure we need this as the most interesting stuff is happening on smart contracts

**Reply by damip** [reply]
> Now, we could also decide to allow domain name keys in the native ledger to allow account abstraction of native coins, but I'm not sure we need this as the most interesting stuff is happening on smart contracts

This would create a lot of side-effects. Let's keep it simple

**Reply by AurelienFT** [reply]
> This would create a lot of side-effects. Let's keep it simple

For the end user it means that he can have fungible tokens on his domain name but not massa coins. Seems strange to have a different behavior between fungible tokens and massa on this no ?

Do you have a non-exhaustive list of the problems it may cause ?

**Reply by sebastien-forestier** [reply]
Same for me, I don't see the side-effects yet, let's discuss them

**Comment by sebastien-forestier** [comment]
Here is a possible V1 allowing Massa Station to proxy .massa addresses, everything else is removed and can be added later:


### State
We add a new table `DNS` to the state.
The `DNS` table is composed of `(.massa string, address)` rows, indexed by the first field (whose items appear at most once).

The first field `.massa string` is the domain name that was booked, and which points to the second field, the `address`.
All rows are immutable.


### Booking a domain name
A new transaction type `DNS_TX` allows to book a domain name.

`DNS_TX(BOOK, dn, address)` books the domain name `dn` and assigns it to `address` (not a .massa). Succeeds only if `dn` is not already booked, and enough coins can be used for storage costs.

As there is no transfer/free possibility, storage cost can be interpreted as a burn.
Booking an address has a cost depending on address length (parameters of the network), with inflated costs (compared to cost of storing a byte) to prevent parking.

### Resolving a domain name with the API
Resolution is done only with the API.

### Storage Costs
Works exactly as for other state tables, paying storage costs when increasing state size.

**Reply by sebastien-forestier** [reply]
V2: add resolving ability: a coin transfer (in a transaction or smart contract) to a .massa address sends coins to the resolved pubkey-hash address or fails.

Optional: A RESOLVE(address) ABI allows to resolve an address from a smart contract. Returns the pubkey-hash or fails.

**Reply by leoloco** [reply]
If the storage is immutable and there is no way to "free" an address, assuming some vulnerability is discovered in some frontend/smart contract/DAO. How could users be alerted of the vulnerability?
